### PR TITLE
Add `SampleTransfer`

### LINKED ISSUE #1390: Untitled story 6132

### PR DESCRIPTION
**Pull request by straight-shoota**
Resolves #1390

I didn't add test cases for creating `SampleTransfer` entries because this logic will be changed in the next iteration #1389. It makes more sense to add specs then.

### CODE REVIEW

File: app/presenters/sample_transfer_presenter.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
+class SampleTransferPresenter
```
Comment (ysbaddaden) [inline app/presenters/sample_transfer_presenter.rb:1]: This presenter looks nice, but this is introducing a new concept into CDx. I'm wondering if it really improves readability. Most methods could probably be on `FileTransfer` directly, but again the presenter looks nice, so :shrug:

@devduarte Any opinion?

File: app/controllers/sample_transfers_controller.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  def index
+    @sample_transfers = SampleTransfer
+      .within(@navigation_context.institution)
+      .includes(:receiver_institution, :sender_institution, sample: [:sample_identifiers])
+
+    @sample_transfers = @sample_transfers.joins(sample: :sample_identifiers).where("sample_identifiers.uuid LIKE concat('%', ?, '%')", params[:sample_id]) unless params[:sample_id].blank?
+    @sample_transfers = @sample_transfers.joins(sample: :batch).where("batches.batch_number LIKE concat('%', ?, '%')", params[:batch_number]) unless params[:batch_number].blank?
+    @sample_transfers = @sample_transfers.joins(:sample).where("samples.isolate_name LIKE concat('%', ?, '%')", params[:isolate_name]) unless params[:isolate_name].blank?
+    @sample_transfers = @sample_transfers.joins(:sample).where("samples.specimen_role = ?", params[:specimen_role]) unless params[:specimen_role].blank?
+
```
[the remark this answers is not in the document]
Comment (ysbaddaden) [reply app/controllers/sample_transfers_controller.rb:11]: An alternative that I use regularly is a nicely formatted `#search(scope)` method on the controller. It's sometimes nicer than cluttering a model with lots of scopes.

```ruby
private

def search(scope)
  if sample_id = params[:sample_id].presence
    scope = scope.joins(...).where(...)
  end

  scope
end
```

**Review by ysbaddaden — APPROVED** [review]
Just a few suggestions to improve readability.

File: app/controllers/sample_transfers_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (straight-shoota) [reply app/controllers/sample_transfers_controller.rb:11]: I totally agree. For now, I just duplicated the logic from `SamplesController` and `BatchesController`. We talked about refactoring common code between samples and batches already in the standup. This is another example for that. I think it makes sense to add this as is now and do a separate refactor.

File: app/presenters/sample_transfer_presenter.rb (hunk 1 above)
Comment (straight-shoota) [reply app/presenters/sample_transfer_presenter.rb:1]: I understand this is a new semantical concept. But it's really just a lightweight PORO.

I don't think these methods can go on the model directly. The purpose of this class is to combine model data with contextual information of the current user and institution.

That's a good use case for a presenter. Think of it as a combination of helper methods with context information.
We could do it without, but that would move all that logic into the template or stand-alone helpers. Both cases lead to more boilerplate and less readable code.

File: app/controllers/sample_transfers_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (diegoliberman) [reply app/controllers/sample_transfers_controller.rb:11]: I created a technical debt work item #1499, to keep this in our radar.

**Comment by devduarte** [comment]
Looks good!

**Review by ysbaddaden — APPROVED** [review]

**Pull request merged**